### ISSUE TITLE
Best approach to domain-specific semantic search?

### ISSUE THREAD

**Issue by shensmobile** [UKPLab/sentence-transformers#1903]
Hi everyone,

I have worked with Transformers for just over a year now for other applications like classification and NER, but want to experiment with semantic search now.  I work in a pretty specific domain (medical) but have access to a lot of sample documents.  I typically work with long documents (easily over 300 tokens) so I expect that I'll need to work with asymmetric search and the MS MARCO dataset.  

I've looked into other issues here on Github and read through a few blog posts and I think that my approach will be to utilize my fine-tuned Bert-based MLM that I use for other downstream tasks that was trained on my medical corpus as a starting point.

I have already done a few rounds of training using the MarginMSE training script for the bi-encoder.  So far, it does appear to work reasonably well for ranking a small test-set of documents, but I know that I will need a cross-encoder to re-rank and I am training that now.

Have people had good results with mixing a pre-trained MLM and the MS MARCO dataset and some unsupervised learning?  I have a large batch of documents that I could use to fine-tune if unsupervised learning would not un-do what it's learning from MS MARCO.  Alternatively, I was thinking about using a text summarization model to create "queries" for each of my documents and creating a dataset in that way.  If I were to do this, should I just out-right replace MS MARCO or use a combination of the two?

Or should I just outright skip all of this and go straight to one of the unsupervised techniques like DeCLUTR?

**Comment by shensmobile** [comment]
Thanks for that heads up.  I decided to just try GPL last night and used the default settings with a sample corpus of over 300k reports.  It should be done in about 6 hours.  Fingers crossed that it improves the performance of my model.  

I didn't have much time to look into the underlying train() code, but by default it looked like GPL wants to train a dot product model so I resorted to using my MarginMSE trained model.  I wonder if I can get it to train using cos_sim instead and feed in my MNRL model.  I feel significantly more confident using that model.

**Comment by HenryL27** [comment]
The tricky thing is that the generated labels come from a cross-encoder, so there is no guarantee on how big they can be. What I've found works decently well is to rescale them to [-1,1] (there's a rescale_range param for gpl.train(); I think it's just an arithmetic normalization). That way the cosine can actually hit the target values. Best of luck!

**Comment by shensmobile** [comment]
If my results are good, I'll be sure to leave a final note here on my adventures with domain adaptation so others can save themselves the headache (although research is always healthy!)

**Comment by mpizosdim** [comment]
Could you share your final note? :)

**Comment by xehu** [comment]
@shensmobile +1 --- I am working on this exact problem (wanting to fine-tune SBERT to have domain adaptation), and would love additional resources.

**Comment by shensmobile** [comment]
Hey everyone,

I think that my edit to my above comment captured most of my final journey, but I know I wrote a lot more on reddit.  Unfortunately the subs that I post on have gone private so I can't see my own comments!

If I recall correctly, my overall path to success was:

1) Train a MLM on my corpus of data.
2) Use the MS MARCO dataset to train a sentence embedding model using the MarginMSE training script and also trained a cross-encoder using the same training script from the MS MARCO repo.  Even though the MS MARCO dataset is not domain specific, it's a stepping stone to building some comparitive knowledge.
3) Take my corpus of data from step 1 and throw it into GPL.  I mimicked the corpus.jsonl structure from the original GPL repo.
4) GPL will spit out a new massive batch of labels that can be used again to retrain the MarginMSE model.  

I took the end result into FAISS and created a mini-demo of a semantic search engine.  This is certainly not the end, there's a LOT of improvements that can be made.  I really hope to be able to contribute more to this space once I have some more free time on my hands again.  I think at this point, I would want to go through the GPL labels and devise a method to reject obviously incorrectly summarized labels for data.  Then I think I would get even better performance.

**Comment by lppier** [comment]
How much data did u use to achieve a decent result? Curious..
